src/Adapter/Product/Image/Repository/ProductImageMultiShopRepository.php

```diff
@@ -0,0 +1,241 @@
+            ;
+        } else {
+            $qb->from($this->dbPrefix . 'image_shop', 'i')
+                ->andWhere('i.id_shop = :shopId')
+                ->setParameter('shopId', $shopConstraint->getShopId()->getValue())
+            ;
+        }
+
+        $results = $qb->andWhere('i.id_product = :productId')
+            ->setParameter('productId', $productId->getValue())
+            ->addOrderBy('i.id_image', 'ASC')
+            ->execute()
+            ->fetchAll()
+        ;
+
+        if (!$results) {
+            return [];
+        }
+
+        $imagesIds = [];
+        foreach ($results as $result) {
+            $imagesIds[] = new ImageId((int) $result['id_image']);
+        }
+
+        return $imagesIds;
+    }
+
+    /**
+     * @param ImageId $imageId
+     *
+     * @return Image
+     *
+     * @throws CoreException
+     */
+    public function get(ImageId $imageId, ShopId $shopId): Image
+    {
+        /** @var Image $image */
+        $image = $this->getObjectModelForShop(
+            $imageId->getValue(),
+            Image::class,
+            ProductImageNotFoundException::class,
+            $shopId
+        );
+
+        return $image;
+    }
+
+    /**
+     * @param ImageId $imageId
+     *
+     * @return ShopId[]
+     */
+    public function getAssociatedShopIds(ImageId $imageId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('id_shop')
+            ->from($this->dbPrefix . 'image_shop')
+            ->where('id_image = :imageId')
+            ->setParameter('imageId', $imageId->getValue())
+        ;
+
+        $result = $qb->execute()->fetchAll();
+        if (empty($result)) {
+            return [];
+        }
+
+        $shops = [];
+        foreach ($result as $shop) {
+            $shops[] = new ShopId((int) $shop['id_shop']);
+        }
+
+        return $shops;
+    }
+
+    public function create(ProductId $productId, ShopConstraint $shopConstraint): Image
+    {
+        $productIdValue = $productId->getValue();
+        $image = new Image();
+        $image->id_product = $productIdValue;
+        $image->cover = !Image::getCover($productIdValue);
+
+        $shopIds = $this->productMultiShopRepository->getShopIdsByConstraint($productId, $shopConstraint);
+        $this->addObjectModelToShops($image, $shopIds, CannotAddProductImageException::class);
+
+        return $image;
+    }
+
+    /**
+     * @param Image $image
+     * @param ShopId[] $shopsToRemoveImageFrom
+     *
+     * @return void
+     */
+    public function deleteFromShops(Image $image, array $shopsToRemoveImageFrom)
+    {
+        $this->deleteObjectModelFromShops($image, $shopsToRemoveImageFrom, CannotDeleteProductImageException::class);
+    }
+
+    /**
+     * @param ImageId $imageId
+     *
+     * @return ShopId[]
+     */
+    public function getShopIdsCoveredBy(ImageId $imageId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
```

Comment:
also minor note (I know we do this way alot), but this can also be simplified to
```suggestion
        $qb = $this->connection->createQueryBuilder()     
```